edx_rbac/utils.py

```diff
@@ -0,0 +1,102 @@
+from edx_rest_framework_extensions.auth.jwt.authentication import JwtAuthentication
+
+from django.conf import settings
+
+
+def get_request_or_stub():
+    """
+    Return the current request or a stub request.
+    If called outside the context of a request, construct a fake
+    request that can be used to build an absolute URI.
+    This is useful in cases where we need to pass in a request object
+    but don't have an active request (for example, in tests, celery tasks, and XBlocks).
+    """
+    request = crum.get_current_request()
+
+    if request is None:
+
+        # The settings SITE_NAME may contain a port number, so we need to
+        # parse the full URL.
+        full_url = "http://{site_name}".format(site_name=settings.SITE_NAME)
+        parsed_url = urlparse(full_url)
+
+        # Construct the fake request.  This can be used to construct absolute
+        # URIs to other paths.
+        return RequestFactory(
+            SERVER_NAME=parsed_url.hostname,
+            SERVER_PORT=parsed_url.port or 80,
+        ).get("/")
+
+    else:
+        return request
+
+
+def get_decoded_jwt_from_request(request):
+    """
+    Grabs jwt from request if possible.
+
+    Returns a decoded jwt dict if it finds it.
+    Returns a None if it does not.
+    """
+    jwt_cookie_name = settings.JWT_AUTH.get('JWT_AUTH_COOKIE') or 'edx-jwt-cookie'
+    jwt_cookie = request.COOKIES.get(jwt_cookie_name, None)
+    
+    if not jwt_cookie:
+        return None
+    return jwt_decode_handler(jwt_cookie)
+
+
+def request_user_has_implicit_access_via_jwt(decoded_jwt, role_name):
+    """
+    Check the request's user access by mapping user's roles found in jwt
+    to local feature roles.
+
+    decoded_jwt is a dict
+    role_name is a string
+
+    Returns a boolean.
+
+    Mapping should be in settings and look like:
+
+        SYSTEM_TO_FEATURE_ROLE_MAPPING = {
+            'enterprise_admin': ['coupon-management', 'data_api_access'],
+            'enterprise_leaner': [],
+            'portal_admin_no_coupons': ['data_api_access']
+        }
+    """
+    jwt_roles_claim = decoded_jwt.get('roles', [])
+
+    feature_roles = []
+    for role_data in jwt_roles_claim:
+        role_in_jwt, object_type, object_key = role_data.split(':') # split should be more robust because of our cousekeys having colons
+        mapped_roles = settings.SYSTEM_TO_FEATURE_ROLE_MAPPING.get(role_in_jwt, [])
+        feature_roles.append(mapped_roles)
+    
+    if role_name in feature_roles:
+        return True
+    return False
+
+
+def user_has_access_via_database(user, role_name, role_class, role_assignment_class):
+    """
+    Checks if there is a role assignment for a given user and role.
+
+    The role object itself is found via the role_name
+    """
+
+    role_obj = role_class.objects.get(name=role_name)
+    try:
+        role_assignment_class.objects.get(user=user, role=role_obj)
```

Comment:
you can get rid of the first line and just do: `role_assignment_class.objects.get(user=user, role__name=role_name)`

Reply:
pushing up in next commit